calendar/page.js

```diff
@@ -464,29 +505,40 @@ function selectRadioButton(value) {
 }
 
 // helper function to build a calendar event object from grist flat record
-function buildCalendarEventObject(record) {
-  let {startDate: start, endDate: end} = record;
-  if (end === null || (end.getTime() <= start.getTime())) {
-    end = thirtyMinutesFrom(start);
+function buildCalendarEventObject(record, colTypes) {
+  let {startDate: start, endDate: end, isAllDay: isAllday} = record;
+  let [startType, endType] = colTypes;
+  endType = endType || startType;
+  start = new calendarHandler.TZDate(start).tz(startType === 'Date' ? 'UTC' : 'Local');
+  end = end ? new calendarHandler.TZDate(end).tz(endType === 'Date' ? 'UTC' : 'Local') : start;
+  if (startType === 'Date' && endType === 'Date') {
+    isAllday = true;
+  }
+  // Workaround for midnight zero-length events not showing up.
+  if (!isAllday && end.valueOf() === start.valueOf() && isZeroTime(end) && isZeroTime(start)) {
+    end = new calendarHandler.TZDate(end).addHours(1);
```

Comment:
I think this is the problem, this 1h is maybe converted by Grist to full next day when event is moved.